pkg/util/graph/util.go

```diff
@@ -0,0 +1,102 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright The KubeVirt Velero Plugin Authors.
+ *
+ */
+
+package vmgraph
+
+import (
+	"github.com/vmware-tanzu/velero/pkg/kuberesource"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/kubevirt-velero-plugin/pkg/util"
+)
+
+// VMObjectGraph represents the graph of objects that can be potentially related to a VirtualMachine
+var VMObjectGraph = map[string]schema.GroupResource{
+	"virtualmachineinstances":           {Group: "kubevirt.io", Resource: "virtualmachineinstances"},
+	"virtualmachineinstancetype":        {Group: "instancetype.kubevirt.io", Resource: "virtualmachineinstancetype"},
+	"virtualmachineclusterinstancetype": {Group: "instancetype.kubevirt.io", Resource: "virtualmachineclusterinstancetype"},
+	"virtualmachinepreference":          {Group: "instancetype.kubevirt.io", Resource: "virtualmachinepreference"},
+	"virtualmachineclusterpreference":   {Group: "instancetype.kubevirt.io", Resource: "virtualmachineclusterpreference"},
+	"datavolumes":                       {Group: "cdi.kubevirt.io", Resource: "datavolumes"},
+	"controllerrevisions":               {Group: "apps", Resource: "controllerrevisions"},
+	"configmaps":                        {Group: "", Resource: "configmaps"},
+	"persistentvolumeclaims":            kuberesource.PersistentVolumeClaims,
+	"serviceaccounts":                   kuberesource.ServiceAccounts,
+	"secrets":                           kuberesource.Secrets,
+	"pods":                              kuberesource.Pods,
+}
+
+func addVeleroResource(name, namespace, resource string, resources []velero.ResourceIdentifier) []velero.ResourceIdentifier {
+	if groupResource, ok := VMObjectGraph[resource]; ok {
+		resources = append(resources, velero.ResourceIdentifier{
+			GroupResource: groupResource,
+			Namespace:     namespace,
+			Name:          name,
+		})
+	}
+	return resources
+}
+
+func addCommonVMIObjectGraph(spec v1.VirtualMachineInstanceSpec, namespace string, resources []velero.ResourceIdentifier) []velero.ResourceIdentifier {
+	resources = addVolumeGraph(spec.Volumes, namespace, resources)
+	resources = addAccessCredentials(spec.AccessCredentials, namespace, resources)
+	return resources
+}
+
+func addVolumeGraph(volumes []v1.Volume, namespace string, resources []velero.ResourceIdentifier) []velero.ResourceIdentifier {
+	for _, volume := range volumes {
+		switch {
+		case volume.DataVolume != nil:
+			resources = addVeleroResource(volume.DataVolume.Name, namespace, "datavolumes", resources)
+			resources = addVeleroResource(volume.DataVolume.Name, namespace, "persistentvolumeclaims", resources)
```

Comment:
@mhenriks do we really need to explicitly backup the pvc if we already handle it in `dv_backup_item_action.go`?

Reply:
Figured out that precisely we want to be explicit as this package may be used by third-party integrations that don't use our plugin's logic.